app/controllers/admin/ex_links_controller.rb

```diff
@@ -0,0 +1,146 @@
+# encoding:UTF-8
+class Admin::ExLinksController < ApplicationController
+  load_permissions_and_authorize_resource
+  before_action :authorize
+  before_action :set_ex_link, only: [:show, :edit, :update, :destroy]
+
+  # GET /ex_links
+  def index
+    if params[:tag]
+      @ex_links = get_exlinks_if_all_tags_present(params[:tag])
+    else
+      @ex_links = ExLink.all
+    end
+    @tags = get_tags
+  end
+
+  # GET /ex_links/1
+  def show
+  end
+
+  # GET /ex_links/new
+  def new
+    @ex_link = ExLink.new
+  end
+
+  # GET /ex_links/1/edit
+  def edit
+  end
+
+  # POST /ex_links
+  def create
+    @ex_link = ExLink.new(ex_link_params)
+    if @ex_link.save
+      redirect_to admin_ex_link_path(@ex_link), notice: alert_create(ExLink)
+    else
+      render :new
+    end
+  end
+
+  # PATCH/PUT /ex_links/1
+  def update
+    if @ex_link.update(ex_link_params)
+      redirect_to admin_ex_link_path(@ex_link), notice: alert_update(ExLink)
+    else
+      render :edit
+    end
+  end
+
+  # DELETE /ex_links/1
+  def destroy
+    @ex_link.destroy
+    redirect_to admin_ex_links_path, notice: alert_destroy(ExLink)
+  end
+
+  # QUERY_BY_TAGS
+
+  # get all ExLinks with specific tag
+  def get_exlinks_if_tag_present(tagname)
+    ExLink.includes(:tags).where(tags: { tagname: tagname })
+  end
+
+  # break string with tags to words and lowercase them
+  # then find appropriate ID of tags with those names
+  def find_tag_ids_by_names(tags_string)
+    ids_of_wanted_tags = []
+    tags_string.split(',').each do |word|
+      possible_tag = Tag.where(tagname: word.downcase.strip)
+      if !possible_tag.empty?
+        ids_of_wanted_tags << possible_tag[0].id
+      end
+    end
+    ids_of_wanted_tags
+  end
+
+  # find all exlinks which have all tags present
+  def get_exlinks_if_all_tags_present(tags_string)
```

Comment:
@davidwessman Joins are very powerful, you can use them in this scenario too. You just have to make a join against ex_link_tags for every tag_name you need to satisfy.

```
select e.* from exlinks e
join ex_link_tags t1 on e.id = t1.tag_id and t1.tag_id == '#{FIRST_ID}'
join ex_link_tags t2 on e.id = t2.tag_id and t2.tag_id == '#{SECOND_ID}'
...
```

Because the join will discard any rows which do not satisfy the join criterion, in the end you will have just the desired exlinks.